hypothesis-python/src/hypothesis/searchstrategy/misc.py

```diff
@@ -85,3 +86,69 @@ def calc_is_cacheable(self, recur):
 
     def do_draw(self, data):
         return d.choice(data, self.elements)
+
+    def do_filtered_draw(self, data, filter):
+        # Set of indices that have been tried so far, so that we never test
+        # the same element twice during a draw.
+        known_bad_indices = set()
+
+        def check_index(i):
+            """Return ``True`` if the element at ``i`` satisfies the filter
+            condition.
+            """
+            if i in known_bad_indices:
+                return False
+            ok = filter.condition(self.elements[i])
+            if not ok:
+                if not known_bad_indices:
+                    filter.note_retried(data)
+                known_bad_indices.add(i)
+            return ok
+
+        # Start with ordinary rejection sampling. It's fast if it works, and
+        # if it doesn't work then it was only a small amount of overhead.
+        for _ in hrange(3):
+            i = d.integer_range(data, 0, len(self.elements) - 1)
+            if check_index(i):
+                return self.elements[i]
+
+        # Arbitrary cutoff to prevent us from wasting too much time on very
+        # large lists, and to avoid a mismatch between block size and index
+        # size when integer_range switches to wide-range generation.
```

Comment:
Probably worth noting in the comment that the latter only kicks in at 24 bits, so time-wasting is the primary driver.

Instead of returning early, we could just stop iterating at the `2**16`th element even if we've not reached `speculative_index`.  I think it's reasonable to only choose one of the higher elements if we can do so in the initial try-three phase, and I'd prefer to bias towards earlier elements than fail to generate.

Reply:
My original motivation was to fix the size-mismatch bug, though I agree that time-wasting is the bigger concern in practice.

I'm also open to the possibility of bikeshedding the cutoff down to a much lower number, since `2 ** 16` was mainly motivated by `2 ** 27` being really slow in coverage tests.

Reply:
I thought about adding a more gradual cutoff (such as scanning the first `n` elements), but in the end I wanted to get the common-case improvement merged without going down too much of an edge-case rabbit hole.

So I'm open to doing better as a follow-up, but I think the simple cutoff is fine for this PR. At the very least it's no worse than what we were doing before.

Reply:
Say 10K then?

And then we can delete the early return block, though IMO we should keep a comment noting that the size cap is for block size as well as performance.

Reply:
Oh, do you mean capping `max_allowed_indices` at 10k?

That sounds simpler and better than what I was imagining.

Reply:
Yep, so we iterate over `for i in hrange(min(len(self.elements), 10 **5)):`

Reply:
Note that as a side-effect of this change, `last_block_length` is no longer usable here.

The write length is now determined entirely by `bit_length(len(self.elements))`.